### PR TITLE
Split download stats on promulgation.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1160/
Test PASSed.

File: internal/charmstore/stats.go [hunk 1]
```diff
@@ -520,7 +520,13 @@ func (s *Store) ArchiveDownloadCounts(id *charm.Reference, refresh bool) (thisRe
 
 func (s *Store) statsCacheFetch(id *charm.Reference) (interface{}, error) {
 	prefix := id.Revision == -1
-	counts, err := s.aggregateStats(EntityStatsKey(id, params.StatsArchiveDownload), prefix)
+	var kind string
```
Comment (mhilton) [inline internal/charmstore/stats.go]: perhaps

``` go
kind := params.StatsArchiveDownload
if id.User == "" {
    kind = params.StatsArchiveDownloadPromulgated
}
```

File: internal/charmstore/stats_test.go [hunk 2]
```diff
@@ -686,7 +686,13 @@ func (s *StatsSuite) TestArchiveDownloadCounts(c *gc.C) {
 }
 
 func setDownloadCounts(c *gc.C, s *charmstore.Store, id *charm.Reference, t time.Time, n int) {
-	key := charmstore.EntityStatsKey(id, params.StatsArchiveDownload)
+	var kind string
```
Comment (mhilton) [inline internal/charmstore/stats_test.go]: same as above

**Comment by mhilton** [comment]
:+1: LGTM with a small suggestion.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1161/
Test PASSed.

File: internal/charmstore/stats.go (hunk 1 above)
Comment (fabricematrat) [inline internal/charmstore/stats.go]: Done.

File: internal/charmstore/stats_test.go (hunk 2 above)
Comment (fabricematrat) [inline internal/charmstore/stats_test.go]: Done.

**Comment by rogpeppe** [comment]
LGTM but I'd like to see a test that specifically tests the failure we had - that a download of a promulgated charm doesn't cause /stats/downloaded:\* to increment by more than one (and verify that the test fails without the changes in this PR).

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1162/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1163/
Test PASSed.

**Comment by fabricematrat** [comment]
I added the test for promulgated entity.

:shipit:  thanks @mhilton  && @rogpeppe

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**